src/backend/distributed/commands/trigger.c

```diff
@@ -203,45 +255,602 @@ get_relation_trigger_oid_compat(HeapTuple heapTuple)
+		EnsureDependenciesExistOnAllNodes(&objectAddress);
+
+		char *triggerName = createTriggerStmt->trigname;
+		return CitusLocalTableTriggerCommandDDLJob(relationId, triggerName,
+												   queryString);
+	}
+
+	return NIL;
+}
+
+
+/*
+ * CreateTriggerStmtObjectAddress finds the ObjectAddress for the trigger that
+ * is created by given CreateTriggerStmt. If missingOk is false and if trigger
+ * does not exist, then it errors out.
  *
- * Note that internal triggers that are created implicitly by postgres for
- * foreign key validation already wouldn't be executed via process utility,
- * hence there is no need to check that case here.
+ * Never returns NULL, but the objid in the address can be invalid if missingOk
+ * was set to true.
  */
-void
+ObjectAddress
+CreateTriggerStmtObjectAddress(Node *node, bool missingOk)
+{
+	CreateTrigStmt *createTriggerStmt = castNode(CreateTrigStmt, node);
+
+	RangeVar *relation = createTriggerStmt->relation;
+	Oid relationId = RangeVarGetRelid(relation, CREATE_TRIGGER_LOCK_MODE, missingOk);
+
+	char *triggerName = createTriggerStmt->trigname;
+	Oid triggerId = get_trigger_oid(relationId, triggerName, missingOk);
+
+	if (triggerId == InvalidOid && missingOk == false)
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
+						errmsg("trigger \"%s\" on relation \"%s\" does not exist",
+							   triggerName, relationName)));
+	}
+
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, TriggerRelationId, triggerId);
+	return address;
+}
+
+
+/*
+ * IsCreateCitusTruncateTriggerStmt returns true if given createTriggerStmt
+ * creates citus_truncate_trigger.
+ */
+static bool
+IsCreateCitusTruncateTriggerStmt(CreateTrigStmt *createTriggerStmt)
+{
+	List *functionNameList = createTriggerStmt->funcname;
+	RangeVar *functionRangeVar = makeRangeVarFromNameList(functionNameList);
+	char *functionName = functionRangeVar->relname;
+	if (strncmp(functionName, CITUS_TRUNCATE_TRIGGER_NAME, NAMEDATALEN) == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * ErrorIfUnsupportedCreateTriggerCommand errors out for unsupported
+ * "CREATE TRIGGER triggerName ON relationName" commands.
+ */
+static void
 ErrorIfUnsupportedCreateTriggerCommand(CreateTrigStmt *createTriggerStmt)
 {
-	RangeVar *triggerRelation = createTriggerStmt->relation;
+	if (IsCreateCitusTruncateTriggerStmt(createTriggerStmt))
+	{
+		return;
+	}
+
+	RangeVar *relation = createTriggerStmt->relation;
+
+	bool missingOk = false;
+	Oid relationId = RangeVarGetRelid(relation, CREATE_TRIGGER_LOCK_MODE, missingOk);
+
+	if (!IsCitusTable(relationId))
+	{
+		return;
+	}
+
+	EnsureCoordinator();
+	ErrorOutForTriggerCommandIfNotCitusLocalTable(relationId);
+}
+
+
+/*
+ * CreateTriggerEventExtendNames extends relation name and trigger name with
+ * shardId, and sets schema name in given CreateTrigStmt.
+ */
+void
+CreateTriggerEventExtendNames(CreateTrigStmt *createTriggerStmt, char *schemaName,
```

Comment:
As far as I know, we prefer putting extend name functions in their respective files as I see for the other objects in codebase.
I think otherwise, `relay_event_utility` would really be a long file

Reply:
umm, I think as extending names is similar thing for different types, it could be useful to put them in a single file even if that is a long file(I am actually not sure how many we have :D). Like if that file only contains a bunch of functions that look very similar, it should probably be okay. That way we can even find a way to make them smaller. But let's leave it for now, I might take a look at this later as a separate PR.